### ISSUE TITLE
Upgrading to Neo4j 2.1.7 breaks the test suite

### ISSUE THREAD

**Issue by jlirochon** [nigelsmall/py2neo#389]
For some reason tests are broken with Neo4j 2.1.7, even though they are passing with Neo4j 2.1.6 and 2.2.0-M03.

Looks related to Cypher transactions not returning expected uris, but I don't see any reasons for this in Neo4j release notes.

If I can help with this, please give me some hint.

**Comment by nigelsmall** [comment]
Thanks for pointing this out. I haven't run any tests at all with 2.1.7 yet. If you are able to find any more detail as to the nature of the error, it'll give me a head start :-)

**Comment by jlirochon** [comment]
Hi Nigel,

The first failing test is `test/core/1-main/path_test.py::TestCreatePath::test_can_create_path` :

        def hydrate(self, data):
            """ Hydrate a dictionary of data to produce a :class:`.Node`,
                :class:`.Relationship` or other graph object instance. The
                data structure and values expected are those produced by the
                `REST API <http://neo4j.com/docs/stable/rest-api.html>`__.
    
                :arg data: dictionary of data to hydrate
    
                """
            if isinstance(data, dict):
                if "self" in data:
                    if "type" in data:
                        return Relationship.hydrate(data)
                    else:
                        return Node.hydrate(data)
                elif "nodes" in data and "relationships" in data:
                    if "directions" not in data:
                        from py2neo.batch import Job, Target
                        node_uris = data["nodes"]
                        relationship_uris = data["relationships"]
                        jobs = [Job("GET", Target(uri)) for uri in relationship_uris]
                        directions = []
                        for i, result in enumerate(self.batch.submit(jobs)):
                            rel_data = result.content
    >                       start = rel_data["start"]
    E                       TypeError: 'NoneType' object is not subscriptable

**Comment by jlirochon** [comment]
After a closer look, here is what I get with 2.1.6 :

    ---------------------------- Captured stdout setup -----------------------------
    > GET http://localhost:7474/
    < 200 OK [100]
    ----------------------------- Captured stdout call -----------------------------
    > GET http://localhost:7474/db/data/
    < 200 OK [732]
    begin
    append 'CREATE p=(n0 {p0})-[r0:`KNOWS`]->(n1 {p1})\nRETURN p' {'p0': {'name': 'Alice'}, 'p1': {'name': 'Bob'}}
    commit
    > POST http://localhost:7474/db/data/transaction/commit [172]
    < 200 OK [chunked]
    [
	    {
	    	'length': 1, 
    		'start': 'http://localhost:7474/db/data/node/6',
    		'end': 'http://localhost:7474/db/data/node/7', 
    		'nodes': [
    			'http://localhost:7474/db/data/node/6', 
	    		'http://localhost:7474/db/data/node/7'
	    	], 
	    	'relationships': [
	    		'http://localhost:7474/db/data/relationship/3'
    		]
    	}
    ]
    > POST http://localhost:7474/db/data/batch [77]
    < 200 OK [454]

And here is what I get with 2.1.7 :

    ---------------------------- Captured stdout setup -----------------------------
    > GET http://localhost:7474/
    < 200 OK [100]
    ----------------------------- Captured stdout call -----------------------------
    > GET http://localhost:7474/db/data/
    < 200 OK [732]
    begin
    append 'CREATE p=(n0 {p0})-[r0:`KNOWS`]->(n1 {p1})\nRETURN p' {'p0': {'name': 'Alice'}, 'p1': {'name': 'Bob'}}
    commit
    > POST http://localhost:7474/db/data/transaction/commit [172]
    < 200 OK [451]
    [
    	{
    		'length': 1, 
    		'start': 'http://localhost:7474/db/data/transaction/commitdb/data/node/0', 
    		'end': 'http://localhost:7474/db/data/transaction/commitdb/data/node/1', 
    		'nodes': [
    			'http://localhost:7474/db/data/transaction/commitdb/data/node/0',
    			'http://localhost:7474/db/data/transaction/commitdb/data/node/1'
    		],
    		'relationships': [
    			'http://localhost:7474/db/data/transaction/commitdb/data/relationship/0'
    		]
    	}
    ]
    > POST http://localhost:7474/db/data/batch [103]
    < 200 OK [137]

As you can see, the uris returned by POST to /db/data/transaction/commit looks weird. Looks like a regression in Neo4j ?

**Comment by nigelsmall** [comment]
That's great info, thanks. Leave it with me.

**Comment by jlirochon** [comment]
Ok, tell me if I can do anything to help

**Comment by jakewins** [comment]
@jlirochon thanks for reporting this, tracking a fix for it here: https://github.com/neo4j/neo4j/pull/3968

**Comment by nigelsmall** [comment]
Issue identified as server regression therefore closing py2neo issue.

**Comment by jeremykendall** [comment]
Hey there,

I'm the CTO and Lead Dev at @GraphStory. We're planning on upgrading to Neo4j 2.1.7, but the plan is on hold until I can confirm this bug and a fix (which I know might have to wait on 2.1.8).

I'm attempting to reverse-engineer @jlirochon's 2.1.7 [test script](https://github.com/nigelsmall/py2neo/issues/389#issuecomment-73429765), but my python-fu is weak. I've reviewed the py2neo documentation, and I'm close, but haven't been able to successfully recreate the test.

Here's what I have so far (works, doesn't output the JSON I want to inspect):

``` python
from py2neo import Graph
from py2neo import watch

watch("httpstream")
watch("py2neo.cypher")

graph = Graph("https://user:password@neo-fake-url.do-stories.graphstory.com:7473/db/data/")

statement = "MERGE (p:Person { name: 'Arthur Dent' }) RETURN p"
tx = graph.cypher.begin()
tx.append(statement)
tx.commit()
```

Here's the ouput:

``` shell
$ python bug-test.py

> GET https://user:password@neo-fake-url.do-stories.graphstory.com:7473/db/data/
< 200 OK [2912]
begin
> GET https://neo-fake-url.do-stories.graphstory.com:7473/
< 200 OK [188]
append u"MERGE (p:Person { name: 'Arthur Dent' }) RETURN p" {}
commit
> POST https://neo-fake-url.do-stories.graphstory.com:7473/db/data/transaction/commit [128]
< 200 OK [chunked]
> GET https://neo-fake-url.do-stories.graphstory.com:7473/db/data/
< 200 OK [2912]
result [u'p'] 1
```

Environment:

* OSX 10.10.2
* Python 2.7.9
* py2neo 2.0.4

``` shell
$ pip show py2neo
---
Name: py2neo
Version: 2.0.4
Location: /usr/local/lib/python2.7/site-packages
Requires:
```

Could I get some tips regarding what I need to do to run my test properly? Thanks in advance.

**Comment by nigelsmall** [comment]
Grabbing the JSON is a bit "internal" so I haven't documented it explicitly. However, you should be able to use `.resource.metadata` on any py2neo object, such as a `Graph` or `Node` and see the underlying JSON.

If you need to go over in any more detail, feel free to drop me and email (nigel at py2neo dot org).

**Comment by jeremykendall** [comment]
Thanks! I don't feel so bad about my python anymore.

**Comment by nigelsmall** [comment]
:-)
